linkerd/app/outbound/src/discover.rs

```diff
@@ -53,11 +54,145 @@ impl<N> Outbound<N> {
+                        } else if let Some(profiles::LogicalAddr(ref addr)) = profile.addr {
+                            policy::BackendDispatcher::BalanceP2c(
+                                ewma,
+                                policy::EndpointDiscovery::DestinationGet {
+                                    path: addr.to_string(),
+                                },
+                            )
+                        } else {
+                            policy::BackendDispatcher::Forward(addr.0, Default::default())
+                        };
+                        policy::ClientPolicy::from_backend(
+                            detect_timeout,
+                            default_queue,
+                            dispatcher,
+                        )
+                    };
+
+                    let policy = policy_from_profile(&*profile_rx.borrow_and_update());
+                    tracing::debug!(
+                        ?policy,
+                        "Policy not found; synthesizing policy from profile..."
+                    );
+                    let (tx, policy_rx) = watch::channel(policy);
+                    tokio::spawn(async move {
+                        loop {
+                            if profile_rx.changed().await.is_err() {
+                                tracing::debug!("Profile watch closed, terminating");
+                                return;
+                            }
+
+                            let policy = policy_from_profile(&*profile_rx.borrow_and_update());
+
+                            tracing::debug!(
+                                ?policy,
+                                "Profile updated; synthesizing policy from profile..."
+                            );
+                            if tx.send(policy).is_err() {
+                                tracing::debug!("Policy watch closed, terminating");
+                                return;
+                            }
+                        }
+                    });
+                    return Ok((Some(profile), policy_rx));
+                }
+
+                let policy = policy::ClientPolicy::from_backend(
+                    detect_timeout,
+                    default_queue,
+                    policy::BackendDispatcher::Forward(addr.0, Default::default()),
+                );
+                tracing::debug!(
+                    ?policy,
+                    "Policy not found; no profile: synthesizing policy from original destination address"
+                );
+
+                let (tx, policy_rx) = watch::channel(policy);
+                // keep the watch channel alive as long as it's in use.
+                tokio::spawn(async move { tx.closed().await });
+                return Ok((None, policy_rx));
+            }
+
+            // Otherwise, errors resolving the policy are considered fatal.
+            Err(policy_error)
+        })
+    })
+}
+
```

Comment:
I think this is a little easier to reason about if we split it up into some smaller chunks: https://github.com/linkerd/linkerd2-proxy/pull/2310